core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java

```diff
@@ -132,430 +112,85 @@ public void applyFailedShards(FailedRerouteAllocation allocation) {
     }
 
-    /**
-     * Return {@code true} if the index is configured to allow shards to be
-     * recovered on any node
-     */
-    private boolean recoverOnAnyNode(@IndexSettings Settings idxSettings) {
-        return IndexMetaData.isOnSharedFilesystem(idxSettings) &&
-                idxSettings.getAsBoolean(IndexMetaData.SETTING_SHARED_FS_ALLOW_RECOVERY_ON_ANY_NODE, false);
-    }
-
-    public boolean allocateUnassigned(RoutingAllocation allocation) {
+    public boolean allocateUnassigned(final RoutingAllocation allocation) {
         boolean changed = false;
-        DiscoveryNodes nodes = allocation.nodes();
-        RoutingNodes routingNodes = allocation.routingNodes();
 
-        // First, handle primaries, they must find a place to be allocated on here
-        final MetaData metaData = routingNodes.metaData();
-        RoutingNodes.UnassignedShards unassigned = routingNodes.unassigned();
+        RoutingNodes.UnassignedShards unassigned = allocation.routingNodes().unassigned();
         unassigned.sort(new PriorityComparator() {
 
             @Override
             protected Settings getIndexSettings(String index) {
-                IndexMetaData indexMetaData = metaData.index(index);
+                IndexMetaData indexMetaData = allocation.metaData().index(index);
                 return indexMetaData.getSettings();
             }
         }); // sort for priority ordering
-        Iterator<ShardRouting> unassignedIterator = unassigned.iterator();
-        while (unassignedIterator.hasNext()) {
-            ShardRouting shard = unassignedIterator.next();
 
-            if (!shard.primary()) {
-                continue;
-            }
-
-            // this is an API allocation, ignore since we know there is no data...
-            if (!routingNodes.routingTable().index(shard.index()).shard(shard.id()).primaryAllocatedPostApi()) {
-                continue;
-            }
-
-            AsyncShardFetch<TransportNodesListGatewayStartedShards.NodeGatewayStartedShards> fetch = asyncFetchStarted.get(shard.shardId());
-            if (fetch == null) {
-                fetch = new InternalAsyncFetch<>(logger, "shard_started", shard.shardId(), startedAction);
-                asyncFetchStarted.put(shard.shardId(), fetch);
-            }
-            AsyncShardFetch.FetchResult<TransportNodesListGatewayStartedShards.NodeGatewayStartedShards> shardState = fetch.fetchData(nodes, metaData, allocation.getIgnoreNodes(shard.shardId()));
-            if (shardState.hasData() == false) {
-                logger.trace("{}: ignoring allocation, still fetching shard started state", shard);
-                unassignedIterator.remove();
-                routingNodes.ignoredUnassigned().add(shard);
-                continue;
-            }
-            shardState.processAllocation(allocation);
+        changed |= primaryShardAllocator.allocateUnassigned(allocation);
+        changed |= replicaShardAllocator.allocateUnassigned(allocation);
```

Comment:
I will do the assert when I remove the primaryAllocated flag in a different change